Modules/hmacmodule.c

```diff
@@ -0,0 +1,1713 @@
+        return 0;
+    }
+
+    int res = 0;
+    Py_BEGIN_ALLOW_THREADS
+        Py_HMAC_HACL_UPDATE(self->state, msg, len, self->name, goto error);
+        goto done;
+#ifndef NDEBUG
+error:
+        res = -1;
+#else
+        Py_UNREACHABLE();
+#endif
+done:
+    Py_END_ALLOW_THREADS
+    return res;
+}
+
+/*[clinic input]
+_hmac.new
+
+    key as keyobj: object
+    msg as msgobj: object(c_default="NULL") = None
+    digestmod as hash_info_ref: object(c_default="NULL") = None
+
+Return a new HMAC object.
+[clinic start generated code]*/
+
+static PyObject *
+_hmac_new_impl(PyObject *module, PyObject *keyobj, PyObject *msgobj,
+               PyObject *hash_info_ref)
+/*[clinic end generated code: output=7c7573a427d58758 input=92fc7c0a00707d42]*/
+{
+    hmacmodule_state *state = get_hmacmodule_state(module);
+    if (hash_info_ref == NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "new() missing 1 required argument 'digestmod'");
+        return NULL;
+    }
+
+    const py_hmac_hinfo *info = find_hash_info(state, hash_info_ref);
+    if (info == NULL) {
+        return NULL;
+    }
+
+    HMACObject *self = PyObject_GC_New(HMACObject, state->hmac_type);
+    if (self == NULL) {
+        return NULL;
+    }
+    HASHLIB_INIT_MUTEX(self);
+    hmac_set_hinfo(state, self, info);
+    int rc;
+    // Create the HACL* internal state with the given key.
+    Py_buffer key;
+    GET_BUFFER_VIEW_OR_ERROR(keyobj, &key, goto error_on_key);
+    rc = hmac_new_initial_state(self, key.buf, key.len);
+    PyBuffer_Release(&key);
+    if (rc < 0) {
+        goto error;
+    }
+    // Feed the internal state the initial message if any.
+    if (msgobj != NULL && msgobj != Py_None) {
+        Py_buffer msg;
+        GET_BUFFER_VIEW_OR_ERROR(msgobj, &msg, goto error);
+        rc = hmac_feed_initial_data(self, msg.buf, msg.len);
+        PyBuffer_Release(&msg);
+#ifndef NDEBUG
+        if (rc < 0) {
+            goto error;
+        }
+#else
+        (void)rc;
+#endif
+    }
+    assert(rc == 0);
+    PyObject_GC_Track(self);
+    return (PyObject *)self;
+
+error_on_key:
+    self->state = NULL;
+error:
+    Py_DECREF(self);
+    return NULL;
+}
+
+/*
+ * Copy HMAC hash information from 'src' to 'out'.
+ */
+static void
+hmac_copy_hinfo(HMACObject *out, const HMACObject *src)
+{
+    assert(src->name != NULL);
+    out->name = Py_NewRef(src->name);
+    assert(src->kind != Py_hmac_kind_hash_unknown);
+    out->kind = src->kind;
+    assert(src->block_size <= Py_hmac_hash_max_block_size);
+    out->block_size = src->block_size;
+    assert(src->digest_size <= Py_hmac_hash_max_digest_size);
+    out->digest_size = src->digest_size;
+    assert(src->api.compute != NULL);
+    assert(src->api.compute_py != NULL);
+    out->api = src->api;
+}
+
+/*
+ * Copy the HMAC internal state from 'src' to 'out'.
+ *
+ * The internal state of 'out' must not already exist.
+ *
+ * Return 0 on success and -1 on failure.
```

Comment:
```suggestion
 * Return 0 on success; otherwise, set an exception and return -1 on failure.
```

I've already edited this locally (and update the other comments).